### PR TITLE
Back to the 🎨  drawing board 🎨 : Revisit object model with singletons in mind

### PR DESCRIPTION
**Pull request by adrianna-chang-shopify**
Closes: https://github.com/adrianna-chang-shopify/ruby-object-model/issues/18 
(hopefully? 🤞  more confusion may ensue 😂 )

As described in the issue above, our existing toyland implementation is confusing because we're relying on _true_ Ruby in some areas, and on our defined toy methods / toy object model in others.

The logic of "what does this `<toy class, ie. ToyClass | ToyModule | ToyObject>` represent in toy land?" (e.g. "what is this class's superclass?") is also bound together with "what should an instance of this `<toy class>` be able to do" (e.g. "a `ToyObject` instance can set and get ivars), which was also confusing.

@tomstuart had the brilliant idea to start with defining these classes in terms of their relationships with one another (at the singleton level, not the instance level). We've also reimplemented our toyland constants as `BasicObjects`, so as to not get confused by the Ruby Class behaviour that gets included automatically by making those things true Ruby classes. TL;DR method / behaviours don't exist until we define them ourselves on these new objects.

The next step will be to reincorporate our existing code into this new model, but for now we're defined everything in a single file. I had to wrap our new code in a module because the tests were unhappy about the redefinition of our `Toy{Object|Module|Class}` constants.

### CODE REVIEW

File: lib/new_impl/class_relationships.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
+module NewImpl
```
Comment (tomstuart) [inline lib/new_impl/class_relationships.rb:1]: At the risk of causing confusion (gulp!), note that we could promote the `Toy` prefix in `ToyObject`, `ToyModule` and `ToyClass` into a namespace by naming this module `Toy` instead of `NewImpl`, giving us `Toy::Object`, `Toy::Module` and `Toy::Class`.

File: test/class_relationships_test.rb [hunk 2]
```diff
@@ -1,77 +1,79 @@
...
+require 'new_impl/class_relationships'
 require 'test_helper'
 
-[
-  [Object, Module, Class],
-  [ToyObject, ToyModule, ToyClass, 'toy_']
-].each do |_Object, _Module, _Class, prefix|
-  describe 'relationships between Object, Module and Class' do
-    include TestHelpers
+module NewImpl
```
Comment (tomstuart) [inline test/class_relationships_test.rb:4]: Putting the whole test inside a module is fine, but the only real advantage is that it allows us to refer to constants defined by that module (`ToyObject`, `ToyModule` and `ToyClass` here) without having to qualify them. Since we only actually refer to those constants once in the entire test, it’s arguably no less convenient to just say `NewImpl::ToyObject`, `NewImpl::ToyModule` and `NewImpl::ToyClass` explicitly and avoid the extra indentation. (If the tests contained a thousand references to those constants then the tradeoff would be different.)

File: test/class_relationships_test.rb [hunk 3]
```diff
@@ -1,77 +1,79 @@
...
+  [
+    [Object, Module, Class],
```
Comment (tomstuart) [inline test/class_relationships_test.rb:6]: Now that the test is inside a module (and, frankly, even if it weren’t, given the mindbending nature of this project) it might be both safer and clearer to fully-qualify these names (`::Object`, `::Module` and `::Class` respectively) to make it completely explicit that we’re referring to the constants defined inside the root namespace versus some other module.

**Review by tomstuart — APPROVED** [review]
Fantastic work, thank you. I feel less confused already! I think.

File: test/class_relationships_test.rb (hunk 3 above)
Comment (tomstuart) [reply test/class_relationships_test.rb:6]: (Trivia: these constants actually live inside `Object`, so we can also refer to them as `::Object::Object`, `::Object::Module` and `::Object::Class` if we wish. 🤯)

File: lib/class_relationships.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  "ToyObject"
+end
+
+def ToyObject.toy_class
+  ToyClass
+end
+
+def ToyObject.toy_kind_of?(klass)
+  superclass = toy_class
+  while superclass
```
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: Ah! Well, in the real Ruby world, I know that `BasicObject` is at the top of the hierarchy, so in that scenario we could stop searching up the ancestor chain once we reach `BasicObject`. In our case, we've established that `ToyObject` is at the top of the hierarchy in our toy model, so presumably once we've reached `ToyObject` we could stop.

Maybe something like this (demonstrated for `ToyObject`, but would be the same for the other two classes):
```ruby
  def ToyObject.toy_kind_of?(klass)
    superclass = toy_class
    while superclass
      return true if klass == superclass
      break if superclass == ToyObject
      superclass = superclass.toy_superclass
    end
    false
  end
```  

We could also add the `klass == superclass` check after the loop, and then check whether we're at `ToyObject` right in the while loop:
```ruby
  def ToyObject.toy_kind_of?(klass)
    superclass = toy_class
    while superclass != ToyObject
      return true if klass == superclass
      superclass = superclass.toy_superclass
    end
    klass == superclass
  end
```

File: lib/class_relationships.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
+# Think of these toy objects as the singletons
+ToyObject = BasicObject.new
+
+def ToyObject.to_s
```
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: I've definitely used `class << self` in Ruby classes before to reopen the singleton class and define class-level methods, like this:
```ruby
class Foo
  class << self 
    def my_singleton_method
    end
  end
end
```
Is it possible to make this work with our `BasicObject` classes? 😄  Or is there another approach altogether that I have not yet seen?!

File: lib/class_relationships.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+def ToyObject.toy_class
+  ToyClass
+end
+
+def ToyObject.toy_kind_of?(klass)
+  superclass = toy_class
+  while superclass
+    return true if klass == superclass
+    superclass = superclass.toy_superclass
```
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: NEAT!! Ruby trivia is my fave ❤️  I got curious about this and was playing around, and realized that we could rewrite `#toy_kind_of` like this:
```ruby
  def ToyObject.toy_kind_of?(klass)
    superclass = toy_class
    Enumerator.produce(superclass) { |klass| superclass = klass.superclass or raise StopIteration }
  end
```
And this even handles cases where our `return true if klass == superclass` line would break (as discussed [here](https://github.com/adrianna-chang-shopify/ruby-object-model/pull/19#discussion_r752079349)), because we raise `StopIteration` as soon as we can no longer do `klass.superclass` (which we can't for `ToyObject`). Cool!

File: test/class_relationships_test.rb (hunk 3 above)
Comment (adrianna-chang-shopify) [reply test/class_relationships_test.rb:6]: Oof, that's a bit mind-bendy 😂

File: lib/new_impl/class_relationships.rb (hunk 1 above)
Comment (adrianna-chang-shopify) [reply lib/new_impl/class_relationships.rb:1]: Woah, very cool! TBH I mostly intended the namespacing module to be a standin while we work on integrating our original classes into this new approach, so I was envisioning this going away soon. I'll address your comment below and fully qualify them though!

File: lib/class_relationships.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: NEAT!! Ruby trivia is my fave :heart:  I got curious about this and was playing around, and realized that we could rewrite `#toy_kind_of` like this:   
```ruby
  def ToyObject.toy_kind_of?(klass)
    superclass = toy_class
    e = Enumerator.produce(superclass) do |klass|
      raise StopIteration if klass == ToyObject
      superclass = klass.toy_superclass
    end
    e.include?(klass)
end
```

File: lib/class_relationships.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: Maybe we ought to write a test that would (currently) fail, something like.
```ruby
 specify 'Object is not kind of other type' do
  AnotherType = BasicObject.new
  refute call_method(_Object, "toy_", :kind_of?, AnotherType)
end
```

[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: > ```ruby
> break if superclass == ToyObject
> ```

This is definitely a valid solution! The downside is that we’re hardcoding some knowledge about a specific object — “`ToyObject` is special” — into all the implementations of `#toy_kind_of?`, which works fine but might make it more difficult to understand & maintain this code in the hypothetical future.

As Sandi Metz is fond of saying: depend upon abstractions, not concretions. What behavioural aspect of `ToyObject` makes it special here, and how do we check for that aspect rather than baking in an assumption about a specific singleton?

[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: > ```ruby
> specify 'Object is not kind of other type' do
>   AnotherType = BasicObject.new
>   refute call_method(_Object, "toy_", :kind_of?, AnotherType)
> end
> ```

That’s a good idea — ultimately I think it’d help us to keep adding these “sense check” unit tests to make sure we’re not missing any edge cases.

(In practice the above test wouldn’t be suitable because the real `#kind_of?` can only accept a class or module as its argument, so we’d want `AnotherType` to be `Class.new` or `Module.new` rather than `BasicObject.new`, but that’s a detail.)

File: lib/class_relationships.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: > Is it possible to make this work with our `BasicObject` classes?

Yes! Note that `self` isn’t special syntax in the `class << self` you wrote, it’s just a different way of referring to `Foo`. So you can just as well write `class << Foo`, or indeed `class << any_object_at_all`, to open up a singleton class for method definitions:

```ruby
>> greeting = 'Hello, world'
=> "Hello, world"
>> class << greeting
     def shout
       upcase + '!!!'
     end
   end
=> :shout
>> greeting.shout
=> "HELLO, WORLD!!!"
>> 'bye'.shout
in `<main>': undefined method `shout' for "bye":String (NoMethodError)
```

File: lib/class_relationships.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: Nice. Can you explain what the reassignment to the `superclass` local variable is doing?

File: lib/class_relationships.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: Actually, I’ve realised I can grasp another opportunity for ✨ barely relevant Ruby trivia ✨ here. Lucky you!

We’re familiar with `self` as always referring to the “current object” within the context where it appears, e.g. inside your `class Foo` example above, the use of `self` refers to `Foo` because that’s the class currently being defined. Arguably the most important role of `self` is to act as the default receiver of messages: if we say `a.b` then that’s explicitly sending the message `:b` to the object `a`, but if we say simply `b` (and `b` isn’t a local variable) then it implicitly means sending the message `:b` to `self`, i.e. `self.b`.

But there’s another piece of invisible context that changes like `self`, which is: what’s the current _module_? Or, by analogy with `self`, what’s the default destination of method definitions? If we say `def a.b` then we’re explicitly defining the method `#b` in `a`’s singleton class, but if we say simply `def b`, in which module are we implicitly placing the definition of `#b`?

This invisible context is hard to talk about because it doesn’t have a concrete name like `self` does, so I guess let’s just call it the “current module”. It’s also hard to even notice that the distinction exists, because the most common case is something like…

```
>> class Foo
     puts "self is #{self}"
     def bar
       # …
     end
   end
self is Foo
=> :bar
>> Foo.instance_methods(false) # exclude inherited methods
=> [:bar]
```

…where the “current object” and “current module” inside `class Foo … end` are the same object, i.e. `Foo`. This is convenient because it makes vanilla class definitions easy to write and understand.

But the two concepts are actually independent, which is most evident in the distinction between [`BasicObject#instance_eval`](https://rubydoc.info/stdlib/core/BasicObject:instance_eval) and [`Module#module_eval`](https://rubydoc.info/stdlib/core/Module:module_eval). Both of these methods evaluate a block, during which `self` will be set to the object you called the method on (i.e. the receiver):

```
>> Foo = Class.new
=> Foo
>> Foo.instance_eval { puts "self is #{self}" }
self is Foo
=> nil
>> Foo.module_eval { puts "self is #{self}" }
self is Foo
=> nil
```

At first glance they appear interchangeable because they do the same thing with respect to `self`, but actually they differ in what value they use as the current module during the block: for `#instance_eval` it’s the receiver’s singleton class, whereas for `#module_eval` it’s the receiver itself (which is always a module, because `#module_eval` is only implemented for modules). So an unqualified `def` puts a method definition in a different module in each case:

```
>> Foo.instance_eval { def a; :a; end }
=> :a
>> Foo.module_eval { def b; :b; end }
=> :b

>> Foo.a
=> :a
>> Foo.new.a
in `<main>': undefined method `a' for #<Foo> (NoMethodError)

>> Foo.new.b
=> :b
>> Foo.b
in `<main>': undefined method `b' for Foo:Class (NoMethodError)

>> Foo.singleton_class.instance_methods(false)
=> [:a]
>> Foo.instance_methods(false)
=> [:b]
```

I suppose the tldr is “`#instance_eval` is yet another way to define methods in an object’s singleton class”. But there’s some deeper knowledge here about the fact that Ruby is tracking this “current module” context at all times and using it to understand which module we’re talking about when we say `def`, and that’s not necessarily (although usually is!) the same object we get when we ask for `self`.

File: lib/class_relationships.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: > What behavioural aspect of ToyObject makes it special here

So the other thought I had when writing this was that we haven't written a `.toy_superclass` method for `ToyObject`, because it doesn't have a superclass! Maybe we could write `ToyObject.toy_superclass` to return `false`, which would instead cause `while superclass` to exit once we've reached ToyObject, because `superclass` would be falsey? I was worried there though that the details of `.toy_kind_of` were dependent on `.toy_superclass` returning a falsey value for `ToyObject`. Is this what you had in mind, or am I still missing the target on this one? 😆

[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: > I think it’d help us to keep adding these “sense check” unit tests

Cool, maybe we can work on this IRL to make sure it makes sense!

[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: > Is this what you had in mind, or am I still missing the target on this one? 😆

Yeah, that’s the direction I was thinking of! What does the real `BasicObject` do?

File: lib/class_relationships.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: I was looking at [these docs](https://rubydoc.info/stdlib/core/Enumerator.produce), and saw that they were reassigning the value to the local variable that was used as the `initial` value in this example:
```ruby
ancestors = Enumerator.produce(node) { |prev| node = prev.parent or raise StopIteration }
```
. In rereading the docs though, I noticed that it says
> The result of the previous iteration is passed to the next one. 

which makes me think reassigning is not necessary at all because `klass.toy_superclass` will just be the result of the block. 😅 I'm wondering if maybe they were reassigning in the example above because the `or raise` portion was following the iteration to the next node?

[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: I think I answered my own question here 😂 After changing `ToyObject.toy_superclass` to return `nil`, I tweaked this to be
```ruby

  def ToyObject.toy_kind_of?(klass)
    superclass = toy_class
    e = Enumerator.produce(superclass) { |klass| klass.toy_superclass or raise StopIteration }
    e.include?(klass)
  end
```
and it works, so I think _even_ with adding `or raise StopIteration` there's no need for reassigning here?

[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: Yeah, I agree. Assigning the next value to a local variable _happens_ to return it from the block, but it’s not necessary here. Neat!

(I hadn’t noticed that the documentation example had an assignment. That’s not very helpful.)

File: lib/class_relationships.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: Looks like `BasicObject.kind_of?` comes from `Kernel`:
```bash
irb(main):003:0> BasicObject.method(:kind_of?)
=> #<Method: #<Class:BasicObject>(Kernel)#kind_of?(_)>
```

With the implementation here:
https://github.com/ruby/ruby/blob/c0c2b31a35e19a47b499b57807bc0a0f9325f6d3/object.c#L789-L798

Which makes use of:
https://github.com/ruby/ruby/blob/c0c2b31a35e19a47b499b57807bc0a0f9325f6d3/object.c#L800-L809

Interestingly, the logic looks familiar! 😉 
```c
    while (cl) {
	if (cl == c || RCLASS_M_TBL(cl) == RCLASS_M_TBL(c))
	    return cl;
	cl = RCLASS_SUPER(cl);
    }
```

So it uses `RCLASS_SUPER` to move up the inheritance chain. I think `RCLASS_SUPER` is aliased to `rb_class_get_superclass` FWICT here:
https://github.com/ruby/ruby/blob/cb4e2cb55a59833fc4f1f6db2f3082d1ffcafc80/include/ruby/internal/core/rclass.h#L46

And [the implementation](https://github.com/ruby/ruby/blob/c0c2b31a35e19a47b499b57807bc0a0f9325f6d3/object.c#L2021-L2064) for `rb_class_get_superclass` helpfully tells us this:
```
 *  Returns nil when the given class does not have a parent class:
 *
 *     BasicObject.superclass   #=> nil
 *
 *--
```

With the implementation like so:
```c
    VALUE super = RCLASS_SUPER(klass);


    if (!super) {
	if (klass == rb_cBasicObject) return Qnil;
	rb_raise(rb_eTypeError, "uninitialized class");
    }
```

So this looks like what we've done here! 🤓

[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: Cool! So it sounds like `#superclass` should return `nil`?

[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply lib/class_relationships.rb]: Yeah! I realized it should actually be `nil` over `false` a little while ago, and pushed the commit for it already :D https://github.com/adrianna-chang-shopify/ruby-object-model/pull/19/commits/a30ada5fc8b5514c56dfa54bee5d68b7cdae219e

[the remark this answers is not in the document]
Comment (tomstuart) [reply lib/class_relationships.rb]: Nice. Way ahead of me! 😂

**Comment by adrianna-chang-shopify** [comment]
@tomstuart I'm going to merge this since I _think_ I got around to all your feedback! 😁  We talked about adding some “sense check” unit tests to make sure we’re not missing any edge cases (specifically WRT `toy_kind_of?` working in situations when X is _not_ a kind of Y), but I think we can address this in a follow up.

Thanks for all the extra Ruby tidbits ❤️
